### PR TITLE
[spaceship] Move endpoint version from hostname to path

### PR DESCRIPTION
**Pull request by lucgrabowski**
### Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.
- [x] I've added or updated relevant unit tests.

### Motivation and Context

There is a need for preparing fastlane for implementation of requests that have `v2` or `v3` version in `path` component as noted in https://github.com/fastlane/fastlane/pull/21890.
There are requests defined in open api spec (https://developer.apple.com/documentation/appstoreconnectapi) that  don't have `v1` in `path`, e.g. https://developer.apple.com/documentation/appstoreconnectapi/read_app_availablity_territories
`GET https://api.appstoreconnect.apple.com/v2/appAvailabilities/{id}/territoryAvailabilities`

From spec:
```
"paths" : {
...
  "/v2/appAvailabilities/{id}/territoryAvailabilities" : {
      "get" : {
...
```

### Description

Endpoint's `hostname`s for `ConnectAPI` had hardcoded endpoint version `v1`, also for `https://api.appstoreconnect.apple.com/v1/` when API token is used.
In this PR `v1` version is moved from `hostname` to paths for 4 `ConnectAPI` clients:
- `tunes_request_client`
- `provisioning_request_client`
- `test_flight_request_client`
- `users_request_client`

Also, this change required change in `stats_middleware` and `stats_middleware_spec` to properly log use of endpoints.

### Testing Steps

Runned unit tests. Not all client methods are tested though, so I checked that all `get`, `post`, `patch` and `delete` requests in code have `v1` in path with simple regular expressions:
```
tunes_request_client\.(get|post|patch|delete)
provisioning_request_client\.(get|post|patch|delete)
test_flight_request_client\.(get|post|patch|delete)
users_request_client\.(get|post|patch|delete)
provisioning_request_client\.(get|post|patch|delete)\("#\{Version::V1\}/
users_request_client\.(get|post|patch|delete)\("#\{Version::V1\}/
test_flight_request_client\.(get|post|patch|delete)\("#\{Version::V1\}/
tunes_request_client\.(get|post|patch|delete)\("#\{Version::V1\}/
```

### CODE REVIEW

**Review by shernade — APPROVED** [review]
There's a lot of modified files, but easy changes. Looks ok to me.

File: spaceship/lib/spaceship/stats_middleware.rb [hunk 1]
```diff
@@ -18,11 +18,11 @@ def services
...
           ServiceOption.new("App Store Connect API (official)", "api.appstoreconnect.apple.com", "JWT"),
           ServiceOption.new("App Store Connect API (web session)", Spaceship::ConnectAPI::TestFlight::Client.hostname.gsub("https://", ""), "Web session"),
-          ServiceOption.new("App Store Connect API (web session)", Spaceship::ConnectAPI::Provisioning::Client.hostname.gsub("https://", ""), "Web session"),
           ServiceOption.new("Legacy iTunesConnect Auth", "idmsa.apple.com", "Web session"),
           ServiceOption.new("Legacy iTunesConnect Auth", "appstoreconnect.apple.com/olympus/v1/", "Web session"),
           ServiceOption.new("Legacy iTunesConnect", Spaceship::TunesClient.hostname.gsub("https://", ""), "Web session"),
-          ServiceOption.new("Legacy iTunesConnect Developer Portal", Spaceship::PortalClient.hostname.gsub("https://", ""), "Web session")
+          ServiceOption.new("Legacy iTunesConnect Developer Portal", Spaceship::PortalClient.hostname.gsub("https://", ""), "Web session"),
+          ServiceOption.new("App Store Connect API (web session)", Spaceship::ConnectAPI::Provisioning::Client.hostname.gsub("https://", ""), "Web session")
```
Comment (shernade) [inline spaceship/lib/spaceship/stats_middleware.rb:25]: why did you change the order in this?

Comment (lucgrabowski) [reply spaceship/lib/spaceship/stats_middleware.rb:25]: I think that the reason was that after the changes
`Spaceship::PortalClient.hostname` is `https://developer.apple.com/services-account/QH65B2`
`Spaceship::ConnectAPI::Provisioning::Client.hostname` is `https://developer.apple.com/services-account/` 
and without changing the order, all the calls would be matched by more generic `https://developer.apple.com/services-account/`

**Review by max-ott — COMMENTED** [review]
Looking good so far. Is the motivation behind this change to make it easier to switch versions (v1 -> v2) for specific models?

**Comment by jakub-przy** [comment]
Hey, have we got any updates on this? Thanks

**Review by max-ott — APPROVED** [review]
Looking good!
